Development/nmos/control_protocol_ws_api.cpp

```diff
@@ -0,0 +1,84 @@
+#include "nmos/control_protocol_ws_api.h"
+
+#include "nmos/slog.h"
+
+namespace nmos
+{
+    // IS-12 Control Protocol WebSocket API
+
+    web::websockets::experimental::listener::validate_handler make_control_protocol_ws_validate_handler(nmos::node_model& model, slog::base_gate& gate_)
+    {
+        return [&model, &gate_](web::http::http_request req)
+        {
+            nmos::ws_api_gate gate(gate_, req.request_uri());
+
+            // RFC 6750 defines two methods of sending bearer access tokens which are applicable to WebSocket
+            // Clients SHOULD use the "Authorization Request Header Field" method.
+            // Clients MAY use a "URI Query Parameter".
+            // See https://tools.ietf.org/html/rfc6750#section-2
+
+            // For now just return true
+            const auto& ws_ncp_path = req.request_uri().path();
+            slog::log<slog::severities::more_info>(gate, SLOG_FLF) << "Validating websocket connection to: " << ws_ncp_path;
+
+            return true;
+        };
+    }
+
+    web::websockets::experimental::listener::open_handler make_control_protocol_ws_open_handler(nmos::node_model& model, nmos::websockets& websockets, slog::base_gate& gate_)
+    {
+        return [&model, &websockets, &gate_](const web::uri& connection_uri, const web::websockets::experimental::listener::connection_id& connection_id)
+        {
+            nmos::ws_api_gate gate(gate_, connection_uri);
+
+            const auto& ws_ncp_path = connection_uri.path();
+            slog::log<slog::severities::info>(gate, SLOG_FLF) << "Opening websocket connection to: " << ws_ncp_path;
+            {
+                // create a websocket connection resource
+
+                nmos::id id = nmos::make_id();
+                websockets.insert({ id, connection_id });
+
```

Comment:
websockets.insert should have be protected by a write lock.

Reply:
Good spot, I agree.

Compare e.g. https://github.com/sony/nmos-cpp/blob/694539a37798a711b38c26386168680b0af99dcb/Development/nmos/events_ws_api.cpp#L58